packages/contracts/test/userStateTransition.test.ts

```diff
@@ -0,0 +1,428 @@
+// @ts-ignore
+import { ethers as hardhatEthers } from 'hardhat'
+import { ethers } from 'ethers'
+import { expect } from 'chai'
+import { genRandomSalt, ZkIdentity } from '@unirep/crypto'
+import { Circuit, NUM_EPOCH_KEY_NONCE_PER_EPOCH } from '@unirep/circuits'
+
+import {
+    genUserStateTransitionCircuitInput,
+    genInputForContract,
+    genNewUserStateTree,
+} from './utils'
+import { Unirep, UserTransitionProof } from '../src'
+import { deployUnirep } from '../src/deploy'
+
+describe('User State Transition', function () {
+    this.timeout(600000)
+    let accounts: ethers.Signer[]
+    let unirepContract: Unirep
+
+    let epoch = 1
+    const user = new ZkIdentity()
+
+    before(async () => {
+        accounts = await hardhatEthers.getSigners()
+
+        unirepContract = await deployUnirep(<ethers.Wallet>accounts[0])
+
+        unirepContract['userSignUp(uint256)'](
+            user.genIdentityCommitment()
+        ).then((t) => t.wait())
+
+        // UST should be performed after epoch transition
+        // Fast-forward epochLength of seconds
+        const epochLength = (
+            await unirepContract.config()
+        ).epochLength.toNumber()
+        await hardhatEthers.provider.send('evm_increaseTime', [epochLength])
+        unirepContract.beginEpochTransition().then((t) => t.wait())
+    })
+
+    it('Valid user state update inputs should work', async () => {
+        const userStateTree = genNewUserStateTree()
+        const reputationRecords = {}
+        const {
+            startTransitionCircuitInputs,
+            processAttestationCircuitInputs,
+            finalTransitionCircuitInputs,
+        } = await genUserStateTransitionCircuitInput(user, epoch, 0, {
+            userStateTree,
+            reputationRecords,
+        })
+
+        // submit start user state tranisiton proof
+        {
+            const { publicSignals, proof } = await genInputForContract(
+                Circuit.startTransition,
+                startTransitionCircuitInputs
+            )
+
+            unirepContract
```

Comment:
All of these need to be awaited. Using `.then` causes a chained promise to be returned, we still need to wait for the chained promise.